format/format.go

```diff
@@ -216,6 +216,9 @@ const (
 	RestoreNameLowercase
 	RestoreNameDoubleQuotes
 	RestoreNameBackQuotes
+
+	RestoreSpacesAroundBinaryOperation
+	RestoreForGeneratedColumn
```

Comment:
What will this flag do? Seems no code is using this flag.

Reply:
`RestoreSpacesAroundBinaryOperation ` in ast/expressions.go, `BinaryOperationExpr`
```go
// Restore implements Node interface.
func (n *BinaryOperationExpr) Restore(ctx *RestoreCtx) error {
	if err := n.L.Restore(ctx); err != nil {
		return errors.Annotate(err, "An error occurred when restore BinaryOperationExpr.L")
	}
	if ctx.Flags.HasSpacesAroundBinaryOperationFlag() {
		ctx.WritePlain(" ")
	}
	if err := n.Op.Restore(ctx); err != nil {
		return errors.Annotate(err, "An error occurred when restore BinaryOperationExpr.Op")
	}
	if ctx.Flags.HasSpacesAroundBinaryOperationFlag() {
		ctx.WritePlain(" ")
	}
	if err := n.R.Restore(ctx); err != nil {
		return errors.Annotate(err, "An error occurred when restore BinaryOperationExpr.R")
	}

	return nil
}
```

`RestoreForGeneratedColumn` in types/field_type.go, `RestoreAsCastType`
```go
		if ctx.Flags.HasForGeneratedColumnFlag() {
			if ft.Charset != charset.CharsetBin {
				ctx.WriteKeyWord(" CHARSET ")
				ctx.WriteKeyWord(ft.Charset)
			}
		} else {
			if ft.Charset != charset.CharsetBin && ft.Charset != mysql.DefaultCharset {
				ctx.WriteKeyWord(" CHARACTER SET ")
				ctx.WriteKeyWord(ft.Charset)
			}
		}
```

Reply:
I see, thanks!

Why don't we always restore as `CHARSET x`? The two are [synonyms](https://dev.mysql.com/doc/refman/8.0/en/charset-syntax.html). And are you sure you don't need the `ft.Charset != mysql.DefaultCharset` check in the `HasForGeneratedColumnFlag()` branch?

Reply:
I try creating a table with `CAST` generated column in MySQL 8. It always says `CHARSET x` in `show create table`.

```sql
mysql> select version();

+-----------+
| version() |
+-----------+
| 8.0.12    |
+-----------+
1 row in set (0.08 sec)

mysql> CREATE TABLE t31 ( `a` int(11) DEFAULT NULL, `b` char(20) GENERATED ALWAYS AS (cast(`a` AS CHAR character set latin1)) VIRTUAL) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
Query OK, 0 rows affected (0.08 sec)

mysql> show create table t31;
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                   |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t31   | CREATE TABLE `t31` (
  `a` int(11) DEFAULT NULL,
  `b` char(20) COLLATE utf8mb4_bin GENERATED ALWAYS AS (cast(`a` as char charset latin1)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> 
mysql> CREATE TABLE t32 ( `a` int(11) DEFAULT NULL, `b` char(20) GENERATED ALWAYS AS (cast(`a` AS CHAR)) VIRTUAL) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
Query OK, 0 rows affected (0.05 sec)

mysql> show create table t32;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                    |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t32   | CREATE TABLE `t32` (
  `a` int(11) DEFAULT NULL,
  `b` char(20) COLLATE utf8mb4_bin GENERATED ALWAYS AS (cast(`a` as char charset utf8mb4)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> 
```

Reply:
@pingyu Yes, I mean could we unify the branches into

```go
		if ft.Charset != charset.CharsetBin && ft.Charset != mysql.DefaultCharset {
			ctx.WriteKeyWord(" CHARSET ")
			ctx.WriteKeyWord(ft.Charset)
		}
```

and thus remove the `RestoreForGeneratedColumn` option?

Reply:
@kennytm Oh, I see...
I tried to remove the `RestoreForGeneratedColumn` option. Pls take a look at the latest commit.
Some UT for `cast` broken and I changed the expected result.
A small incompatibility with MySQL 8 exists, because of `ft.Charset != mysql.DefaultCharset`

MySQL 8:
```sql
mysql> set names utf8mb4;
Query OK, 0 rows affected (0.00 sec)

mysql> CREATE TABLE t42 ( `a` int(11) DEFAULT NULL, `b` char(20) GENERATED ALWAYS AS (cast(`a` AS CHAR)) VIRTUAL) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
Query OK, 0 rows affected (0.05 sec)

mysql> show create table t42;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                    |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t42   | CREATE TABLE `t42` (
  `a` int(11) DEFAULT NULL,
  `b` char(20) COLLATE utf8mb4_bin GENERATED ALWAYS AS (cast(`a` as char charset utf8mb4)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.12    |
+-----------+
1 row in set (0.00 sec)
```

TiDB:
```sql
mysql> CREATE TABLE t42 ( `a` int(11) DEFAULT NULL, `b` char(20) GENERATED ALWAYS AS (cast(`a` AS CHAR)) VIRTUAL) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
Query OK, 0 rows affected (0.01 sec)

mysql> show create table t42;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t42   | CREATE TABLE `t42` (
  `a` int(11) DEFAULT NULL,
  `b` char(20) GENERATED ALWAYS AS (cast(`a` as char)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> 
mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                                                |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta-171-gf20cd87be-dirty
Git Commit Hash: f20cd87bec029a01f27d00d0329c51b2c7c38184
Git Branch: show_create_table_virtual
UTC Build Time: 2019-03-09 03:24:19
GoVersion: go version go1.11.2 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> 
```